File: tuan2/PhanB/bai44.py
import time
import random

def selection_sort(arr):
    a = list(arr)
    n = len(a)
    for i in range(n - 1):
        min_idx = i
        for j in range(i + 1, n):
            if a[j] < a[min_idx]:
                min_idx = j
        a[i], a[min_idx] = a[min_idx], a[i]
    return a

def main():
    sizes = [500, 1000, 2000, 4000]
    times = []
    
    # Khởi động bộ nhớ (warm-up)
    selection_sort([random.randint(1, 100) for _ in range(50)])
    
    for n in sizes:
        min_t = float('inf')
        for _ in range(3):
            data = [random.randint(1, 100000) for _ in range(n)]
            t0 = time.perf_counter()
            selection_sort(data)
            t1 = time.perf_counter()
            min_t = min(min_t, t1 - t0)
        times.append(min_t)
        
    print(f"{'n':<8}{'Thoi gian (s)':<18}{'Ti le T(2n)/T(n)'}")
    print("-" * 40)
    for i in range(len(sizes)):
        ratio = f"{times[i] / times[i-1]:.2f}" if i > 0 else "-"
        print(f"{sizes[i]:<8}{times[i]:<18.6f}{ratio}")

if __name__ == '__main__':
    main()

File: tuan2/PhanB/bai44.jl
using Random
using Printf

function selection_sort(a)
    arr = copy(a)
    n = length(arr)
    for i in 1:(n - 1)
        min_idx = i
        for j in (i + 1):n
            if arr[j] < arr[min_idx]
                min_idx = j
            end
        end
        arr[i], arr[min_idx] = arr[min_idx], arr[i]
    end
    return arr
end

function main()
    sizes = [500, 1000, 2000, 4000]
    times = Float64[]
    
    # Chạy khởi động để biên dịch JIT
    selection_sort(rand(1:100, 50))
    
    for n in sizes
        min_t = Inf
        for _ in 1:3
            data = rand(1:100000, n)
            t = @elapsed selection_sort(data)
            min_t = min(min_t, t)
        end
        push!(times, min_t)
    end
    
    println("n\tThoi gian (s)\tTi le T(2n)/T(n)")
    for i in 1:length(sizes)
        ratio = i > 1 ? @sprintf("%.2f", times[i] / times[i-1]) : "-"
        @printf("%d\t%.6f\t%s\n", sizes[i], times[i], ratio)
    end
end

main()